rfc/rfc-89/rfc-89.md

```diff
@@ -0,0 +1,297 @@
+
+## Design
+### Config
+Add a new config named `hoodie.bucket.index.partition.expressions` default null. Users can specify the bucket numbers for different
+partitions by configuring a JSON expression. For example
+```json
+{
+    "expressions": [
+        {
+            "expression": "11-11",
+            "bucketNumber": 10,
+            "rule": "regex"
+        },
+        {
+            "expression": "01-01",
+            "bucketNumber": 20,
+            "rule": "regex"
+        },
+        {
+            "expression": "dt>2025-01-01",
+            "bucketNumber": 20,
+            "rule": "range"
+        }
+    ]
+}
+```
+for partitions match different rule will get and set corresponding bucket number.
+
+We can determine whether the user is currently using the partition-level bucket index based on the value of
+`hoodie.bucket.index.partition.expressions`. If it is null, the processing behavior will be exactly the same as the current logic.
+The advantage of this approach is that it can be fully compatible with the current design of the table-level bucket index,
+enabling a seamless migration for users without their awareness.
+
+### Hashing Metadata
+The hashing metadata will be persisted as files named as `<instant>.simple_hashing_meta` for current table, it stores in
+`.hoodie/.simple_hashing_meta/` directory and contains the following information in a readable encoding
+
+```avro schema
+{
+   "namespace":"org.apache.hudi.avro.model",
+   "type":"record",
+   "doc": "hashing meta for current table using partition level simple bucket index",
+   "name":"SimpleHashMeta",
+   "fields":[
+     {
+       "name":"version",
+       "type":["int", "null"],
+       "default": 1
+     },
+     {
+       "name": "partitionMeta",
+       "type": [
+         "null", {
+           "type":"map",
+           "values": {
+             "type": "record",
+             "name": "PartitionHashInfo",
+             "fields": [
+               {
+                 "name": "bucketNumber",
+                 "type": ["null", "int"],
+                 "default": null
+               },
+               {
+                 "name": "instant",
+                 "type":["null","string"],
+                 "default": null
+               }
+             ]
+           }}],
+       "default" : null
+     },
+     {
+       "name":"defaultBucketNumber",
+       "type":["int", "null"],
+       "default": 256
+     },
+   ]
+}
+```
+We will write <instant>.simple_hashing_meta during commit action:
+1. Call TransactionManager to beginTransaction
+2. Get last complete T1.simple_hashing_meta
+3. Merging T1.simple_hashing_meta with new written partitions as T2.simple_hashing_meta
+4. Call TransactionManager to endTransaction
+
+### SimpleBucketIdentifier
+```java
+public class SimpleBucketIdentifier extends BucketIdentifier {
+  // use expressionEvaluator to compute bucket number for given partition path
+  private BucketCalculator calculator;
```

Comment:
Do we need to read and cache the partition metadata for partitioner?

Reply:
Yes we need. 
The Bucket Identifier loads the partition bucket count information from the metadata as a cache. 
During the Partitioner and Writer stages, when using the Identifier for shuffling and tagging, it will first attempt to find the bucketNumber of the corresponding partition from the cache. 
If the cache was missed, it will calculate the corresponding value based on the user's expression, and then submit it to the metadata during the commit stage.

Reply:
Cache is not safety for multi-writers, still need to think through the concurrent conflicts for mapping file and the writer cache.

Reply:
`hashing_meta` should be unmodifiable during writing.

Reply:
First, in a multi-write scenario, we expect different jobs to load the same expression configuration file. In this way, the in-memory caches of multiple jobs are always synchronized because the calculation logic for partitions is consistent.

Second, in the scenario where different jobs load different versions of expressions, we will conduct two conflict checks. 
1. The first check is performed in the partitioner, where we compare the bucket numbers in the cache with the calculated results for conflict detection. 
2. The second check is carried out during the submission phase to ensure consistency.

Also as @LinMingQiang said, hashing_meta couldn't be modified during writing.